pkg/operator/ceph/cluster/watcher.go

```diff
@@ -193,3 +197,83 @@ func (c *clientCluster) getCephCluster() *cephv1.CephCluster {
 
 	return &clusterList.Items[0]
 }
+
+// each OSD nearfull status is checked
+// OSD capacity is increased if we get nearfull warning
+func increaseOSDsCapacity(cephStatusChecker *cephStatusChecker) {
+	osdUsage, err := cephclient.GetOSDUsage(cephStatusChecker.context, cephStatusChecker.clusterInfo)
+	if err != nil {
+		logger.Debugf("failed to get osd usage")
+	}
+	// by default nearfull_ratio is set to 85%
+	osdNearfullRatio := 85
+
+	var osdsNearfullList []int
+	for _, osdStatus := range osdUsage.OSDNodes {
+		osdId := osdStatus.ID
+		storageUtilization, err := osdStatus.Utilization.Int64()
+		if err == nil {
+			if storageUtilization > int64(osdNearfullRatio) {
+				osdsNearfullList = append(osdsNearfullList, osdId)
+			}
+		}
+
+	}
+	increaseDeviceSetCapacity(cephStatusChecker, osdsNearfullList)
+}
+
+// list the storageClassDeviceSets that are linked to OSDs, and increase there storageSize
+func increaseDeviceSetCapacity(cephStatusChecker *cephStatusChecker, osdsNearfullList []int) {
+	var storageClassDeviceSetsList = util.NewSet()
+	for _, osdID := range osdsNearfullList {
+		storageClassDeviceSetName, err := getDeviceSetName(cephStatusChecker, osdID)
+		if err != nil {
+			logger.Debugf("unable to fetch deviceSet name %+v", err)
+		}
+		storageClassDeviceSetsList.Add(storageClassDeviceSetName)
+	}
+
+	clusterName := cephStatusChecker.clusterInfo.NamespacedName()
+	cephCluster, err := cephStatusChecker.context.RookClientset.CephV1().CephClusters(clusterName.Namespace).Get(context.TODO(), clusterName.Name, metav1.GetOptions{})
+	if err != nil {
+		logger.Errorf("failed to retrieve ceph cluster %q in namespace %q", clusterName.Name, clusterName.Namespace)
+	} else {
+		for storageClassDeviceSet := range storageClassDeviceSetsList.Iter() {
```

Comment:
`deviceSet` is a nice short name for `storageClassDeviceSet` :)
```suggestion
		for deviceSet := range storageClassDeviceSetsList.Iter() {
```